csc: Add ability to update CSC defaults during upgrade/downgrade

Ebaneck opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->
'salt', 'kubernetes', 'csc'

**Context**: 

See #2488 

**Summary**:

- This PR adds the ability to update default CSC values during upgrade and downgrade while keeping the CSC Configmap intact. This works because we are able to merge the user customizations defined within Configmaps into the defaults packaged alongside each chart and then use the resulting dict object to render & deploy Metalk8s addons.

Flow:
1. Create CSC defaults for each Addon  e.g `alertmanager.yaml`
2. Import the CSC defaults within the charts using the `render.py` script
3. Call a salt module that takes arguments `csc ConfigMap name`, `namespace` and `default csc`
4. The module should merge the configs read in `csc ConfigMap` to those of `default CSC` and use  the resulting object to deploy the Addon


**Acceptance criteria**: 

no regression

---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #2488 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

Ebaneck committed:
salt: render utility should import csc defaults

Since we make use of csc defaults in the charts, the render
utility script should import the csc defaults using jinja syntax

Ebaneck committed:
kubernetes: create empty service configuration configmaps

This commit refactors service configuration configmaps to become
empty objects that will hold and persist user customizable settings

Ebaneck committed:
charts: render the prometheus-operator charts with default service configurations

This chart is re-rendered using:
```
$ ./charts/render.py prometheus-operator --namespace metalk8s-monitoring
  charts/prometheus-operator.yaml --service-config grafana metalk8s-grafana-config
  --service-config prometheus metalk8s-prometheus-config --service-config alertmanager
  metalk8s-alertmanager-config charts/prometheus-operator/ > salt/metalk8s/addons/prometheus-operator/deployed/chart.sls
```

Ebaneck committed:
charts: re-render the dex charts to use default service configurations

This chart is re-rendered using:

```
$ ./charts/render.py dex --namespace metalk8s-auth charts/dex.yaml
  --service-config dex metalk8s-dex-config charts/dex/ > salt/metalk8s/addons/dex/deployed/chart.sls
```

Ebaneck committed:
salt: implement merge strategy to make `csc` update-able

This commit makes use of the salt module `dictupdate.merge` ensuring
that default CSC values are merged with user specified CSC while still
providing a fallback configuation for extreme scenarios.

This above merge is implemented to ensure that we can do the following:
1. Deploy metalk8s addons with minimum default service values
2. Metalk8s administrators can update and add new configurations values
3. In cases of no Administrator configurations, the default values will be used
   to deploy the services.
4. In cases where CSC configmaps are removed, we can fall back to default values

closes: #2488

Ebaneck committed:
csc: add default csc configurations as YAML files

This commit adds csc configuration files for dex, alertmanager,
prometheus and grafana to their respective addon folders.

These default configurations will be read during addon rendering by salt
and then the default values will be used to deploy the respective services.

bert-e commented:
# Hello ebaneck,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2488-make-csc-upgrade-able` into
`development/2.5`
* [w/2.6/improvement/2488-make-csc-upgrade-able](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2488-make-csc-upgrade-able) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

bert-e commented:
# History mismatch



Merge commit #ffa183d799bcbde3875b3e21e20b9e1265188f67 on the integration branch
`w/2.6/improvement/2488-make-csc-upgrade-able` is merging a branch which is neither the current
branch `improvement/2488-make-csc-upgrade-able` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2488-make-csc-upgrade-able` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck committed:
test: refactor csc tests to handle empty csc ConfigMaps

Initially, we expect default CSC ConfigMaps to contain default values and
so we read these default values e.g `replicas` to perform csc persistence tests

But with our new approach, default CSC ConfigMaps are deployed
without default service values, so we need to adapt the test accordingly

bert-e commented:
# History mismatch



Merge commit #9b462f150a35bf52f405426bb1f58e648ab2764a on the integration branch
`w/2.6/improvement/2488-make-csc-upgrade-able` is merging a branch which is neither the current
branch `improvement/2488-make-csc-upgrade-able` nor the development branch
`development/2.6`.

It is likely due to a rebase of the branch `improvement/2488-make-csc-upgrade-able` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

Ebaneck commented on charts/render.py:
```diff
@@ -212,15 +213,27 @@ def main():
+            ) for service_config in args.service_configs
+        )
+    else:
+        import_csc_yaml = ''
+        config = ''
```
should we tolerate `a = b = <value>`

alexandre-allard commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -35,13 +38,19 @@ def get_service_conf(
+            'Expected default CSC for ConfigMap {} but got {}'.format(
+                configmap_name, default_csc
+            )
+        )
```
In a module, you should use `CommandExecutionError` rather than an `Exception`.
Also, for the second `if` I would check that the argument is a `dict` rather than checking if it's evaluated to false or not.
```suggestion
    if not configmap_name:
        raise CommandExecutionError(
            'Expected a ConfigMap name but got {}'.format(configmap_name)
        )
    if not isinstance(default_csc, dict):
        raise CommandExecutionError(
            'Expected default CSC for ConfigMap {} but got {}'.format(
                configmap_name, default_csc
            )
        )
```

Ebaneck replied:
Done

alexandre-allard commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -99,6 +104,14 @@ def get_service_conf(
+        default_csc, config, strategy='recurse',
+        merge_lists=True
+    )
+    if not merged_config:
+        raise CommandExecutionError(
+            'Expected to merge configs {} into {}'.format(
+                config, default_csc
+            )
+        )
```
Is this even possible to have an empty dict (or anything evaluated to False) here ?

Ebaneck replied:
Not sure it is possible to have an empty dict so i will remove

alexandre-allard commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -35,13 +38,19 @@ def get_service_conf(
     .. code-block:: bash
 
-        salt-call metalk8s_service_configuration.get_service_conf "metalk8s-monitoring" "metalk8s-prometheus-config"
+        salt-call metalk8s_service_configuration.get_service_conf "metalk8s-monitoring" "metalk8s-prometheus-config" default_csc
```
Your example should be valid (or at least what's `default_csc`), this is not the case here, we don't really understand what's `default_csc`.

Ebaneck replied:
Done

alexandre-allard commented on salt/_modules/metalk8s_service_configuration.py:
```diff
@@ -27,6 +29,7 @@ def get_service_conf(
     Arguments:
         configmap_name: the ConfigMap name
         namespace: the Namespace where the ConfigMap is stored
+        default_csc: the csc configurations imported from YAML file
```
Should be `CSC`, not `csc`. And isn't `csc configurations` redundant as it stands for `Cluster & Services Configurations` ?

Ebaneck replied:
Let's go with `CSC`
Done

alexandre-allard commented on salt/metalk8s/addons/prometheus-operator/deployed/service-configuration.sls:
```diff
@@ -39,9 +39,16 @@ Create grafana-config ConfigMap:
-              deployment:
-                replicas: 1
+            spec: {}
+        # example configuration
+          # example.yaml: |-
+          #   apiVersion: addons.metalk8s.scality.com
+          #   kind: GrafanaConfig
+          #   spec:
+          #     # Configure the Grafana Deployment
+          #     deployment:
+          #       replicas: 1
```
Not sure we want to keep this here as it does not really help, we should rather have some documentation exposing the CSC defaults.

Ebaneck replied:
Ok, going to open a documentation ticket to track this
Here is the ticket: https://github.com/scality/metalk8s/issues/2522

alexandre-allard commented on salt/metalk8s/addons/prometheus-operator/deployed/service-configuration.sls:
```diff
@@ -63,9 +70,16 @@ Create prometheus-config ConfigMap:
-              deployment:
-                replicas: 1
+            spec: {}
+        # example configuration
+          # example.yaml: |-
+          #   apiVersion: addons.metalk8s.scality.com
+          #   kind: PrometheusConfig
+          #   spec:
+          #     # Configure the Prometheus Deployment
+          #     deployment:
+          #       replicas: 1
+
```
Same as above

alexandre-allard commented on salt/metalk8s/addons/prometheus-operator/deployed/service-configuration.sls:
```diff
@@ -87,27 +101,17 @@ Create alertmanager-config ConfigMap:
-                    - name: 'null'
-                  inhibit_rules: []
+            spec: {}
+        # example configuration
+          # example.yaml: |-
+          #   apiVersion: addons.metalk8s.scality.com
+          #   kind: AlertmanagerConfig
+          #   spec:
+          #     # Configure the Alertmanager Deployment
+          #     deployment:
+          #       replicas: 1
+          #     notification:
+          #       config: {}
```
Same

alexandre-allard commented on salt/metalk8s/addons/dex/deployed/service-configuration.sls:
```diff
@@ -23,18 +23,23 @@ Create dex-config ConfigMap:
-              connectors: []
-
+            spec: {}
+        # example configuration
+          # example.yaml: |-
+          #   apiVersion: addons.metalk8s.scality.com
+          #   kind: DexConfig
+          #   spec:
+          #     # Configure the Dex Deployment
+          #     deployment:
+          #       replicas: 2
+          #     localuserstore:
+          #       enabled: true
+          #       userlist:
+          #       - email: "read_only@metalk8s.invalid"
+          #         hash: "$2a$10$2b2cU8CPhOTaGrs1HRQuAueS7JTT5ZHsHSzYiFPm1leZck7Mc8T4W"
+          #         username: "read_only"
+          #         userID: "89b3684b-db28-b7v3-a99c-j99305v2"
+          #     connectors: []
```
Not sure we want to keep this here as it does not really help, we should rather have some documentation exposing the CSC defaults.

Ebaneck commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/2488-make-csc-upgrade-able` into
`development/2.5`
* [w/2.6/improvement/2488-make-csc-upgrade-able](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/2488-make-csc-upgrade-able) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

alexandre-allard reviewed (approved):
LGTM,

Ebaneck commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`



Please check the status of the associated issue None.

Goodbye ebaneck.
